File: articles/biostatistics-experimental-design-biology/r/power_simulation.R
# Power simulation workflow in R.

data_path <- file.path("data", "power_scenarios.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "power_scenarios.csv")
}

simulate_power <- function(sample_size, effect_size, sigma, n_sim, seed) {
  set.seed(seed)
  significant <- 0

  for (i in seq_len(n_sim)) {
    control <- rnorm(sample_size, mean = 0, sd = sigma)
    treated <- rnorm(sample_size, mean = effect_size, sd = sigma)

    difference <- mean(treated) - mean(control)
    se <- sqrt(var(control) / sample_size + var(treated) / sample_size)

    if (!is.na(se) && se > 0 && abs(difference / se) > 1.96) {
      significant <- significant + 1
    }
  }

  significant / n_sim
}

scenarios <- read.csv(data_path)

scenarios$estimated_power <- mapply(
  simulate_power,
  scenarios$sample_size_per_group,
  scenarios$effect_size,
  scenarios$sigma,
  scenarios$n_sim,
  scenarios$seed
)

print(round(scenarios, 5))

File: articles/biostatistics-experimental-design-biology/python/power_simulation.py
"""
Power simulation workflow.

Run:
    python python/power_simulation.py
"""

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "power_scenarios.csv"


def simulate_power(sample_size: int, effect_size: float, sigma: float, n_sim: int, seed: int) -> float:
    """Approximate two-group power using a z threshold."""

    rng = np.random.default_rng(seed)
    significant = 0

    for _ in range(n_sim):
        control = rng.normal(loc=0.0, scale=sigma, size=sample_size)
        treated = rng.normal(loc=effect_size, scale=sigma, size=sample_size)

        difference = treated.mean() - control.mean()
        se = np.sqrt(control.var(ddof=1) / sample_size + treated.var(ddof=1) / sample_size)

        if se > 0 and abs(difference / se) > 1.96:
            significant += 1

    return significant / n_sim


def main() -> None:
    scenarios = pd.read_csv(DATA_PATH)

    rows = []

    for _, row in scenarios.iterrows():
        rows.append(
            {
                "scenario": row["scenario"],
                "sample_size_per_group": row["sample_size_per_group"],
                "effect_size": row["effect_size"],
                "sigma": row["sigma"],
                "estimated_power": simulate_power(
                    sample_size=int(row["sample_size_per_group"]),
                    effect_size=float(row["effect_size"]),
                    sigma=float(row["sigma"]),
                    n_sim=int(row["n_sim"]),
                    seed=int(row["seed"]),
                ),
            }
        )

    print(pd.DataFrame(rows).round(5).to_string(index=False))


if __name__ == "__main__":
    main()
